feat(editor): Indicate dirty nodes with yellow borders/connectors on canvas

autologie opened the pull request:
## Summary
This PR adds the new feature to highlight edited nodes after an execution with yellow borders/connectors, including its all downstream nodes. This helps user understand which nodes will run when they do a partial execution.

![Screenshot from 2025-02-07 13-10-23](https://github.com/user-attachments/assets/f9d95ab0-1198-4e7c-b219-112a64c510a6)

## Related Linear tickets, Github issues, and Community forum posts
https://linear.app/n8n/issue/SUG-2/feature-changing-dirty-nodes-yellow

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

autologie committed:
feat(editor): indicate dirty nodes with yellow borders/connectors on canvas

autologie committed:
refactorthe function to determine stale nodes and rename symbols

autologie committed:
add a safegard against infinite recursion

autologie committed:
reuse calculated staleness in OutputPanel

autologie committed:
Merge branch 'master' into sug-2-indicate-dirty-nodes-on-canvas

autologie committed:
add feature flag check for highlighting stale nodes

autologie committed:
add/update tooltips

autologie committed:
de-prioritize warning border color

autologie committed:
rename 'stale' to 'dirty'

autologie committed:
add unit tests

autologie committed:
revert describe.only()

autologie committed:
change icon

autologie committed:
fix: if new partial execution is not enabled, nothing should change

autologie committed:
fix: nodes without runData should remain gray

autologie committed:
Merge branch 'master' into sug-2-indicate-dirty-nodes-on-canvas

MiloradFilipovic commented on packages/editor-ui/src/stores/workflows.store.ts:
```diff
@@ -273,6 +274,52 @@ export const useWorkflowsStore = defineStore(STORES.WORKFLOWS, () => {
+		}
+
+		const dirtiness: Record<string, CanvasNodeDirtiness | undefined> = {};
+		const visitedByName: Record<string, true | undefined> = {};
```
Nit: Would using `Set` to track visited nodes here be clearer and type-safer?
```suggestion
		const visitedByName = new Set<string>();
```

MiloradFilipovic reviewed (dismissed):
Works as expected and the code is looking good. I left one really minor comment and here's one more minor thing that we probably want to change:
For disabled nodes on the dirty path, the connection that goes through them is green (I guess we want to update it also to yellow?):
![image](https://github.com/user-attachments/assets/694afbde-ba19-4259-963e-bae8fe7329c2)

autologie commented:
> Works as expected and the code is looking good. I left one really minor comment and here's one more minor thing that we probably want to change: For disabled nodes on the dirty path, the connection that goes through them is green (I guess we want to update it also to yellow?): ![image](https://private-user-images.githubusercontent.com/2598782/411490581-694afbde-ba19-4259-963e-bae8fe7329c2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkxODE0NzQsIm5iZiI6MTczOTE4MTE3NCwicGF0aCI6Ii8yNTk4NzgyLzQxMTQ5MDU4MS02OTRhZmJkZS1iYTE5LTQyNTktOTYzZS1iYWU4ZmU3MzI5YzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTBUMDk1MjU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzQxODgwN2FiOWI0NTU2NjkxMDNhMzZkMjIzMzU1ZjBkN2E0OWNmZDUwNDkyY2QzMjVkZDFhOWI1YzdjNTk0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vypHIswkrH2bMAQkVbjuO4O8Wkw5dxclraGUgZDfhig)

Thank you for the finding! I will update the color to be aligned with the rest.

autologie committed:
refactor: use Set for tracking visited nodes

autologie committed:
fix: node injection/removal and toggling enabled/disabled should change downstream dirtiness

autologie committed:
refactor: use history as the place to store data for calculating dirtiness

autologie committed:
Merge branch 'master' into sug-2-indicate-dirty-nodes-on-canvas

autologie committed:
remove duplicate store

autologie committed:
cover all connection related dirtiness checks

autologie committed:
feat: updating pinned data makes downstream dirty

autologie committed:
revert unrelated changes

autologie committed:
fix failing tests

autologie committed:
fix: reflect dirtiness to strikethrough line

autologie committed:
fix compile error

autologie committed:
refactor

autologie committed:
add unit tests

autologie committed:
add more tests

autologie committed:
make dirty check work for sub nodes

autologie committed:
refactor test

autologie committed:
fix: output should not become dirty just by opening NDV

autologie committed:
fix: should not crash if metadata for node does not exist

autologie committed:
fix: warning icon is missing in stale output panel of an AI model node

autologie committed:
Merge branch 'master' into sug-2-indicate-dirty-nodes-on-canvas

autologie committed:
try different logic for highlighting dirty nodes

autologie committed:
fix: send startNodes for now

autologie committed:
change how dirtiness is shown

autologie committed:
revert the condition to show icon in NDV

autologie committed:
update tooltip text for dirty nodes

autologie committed:
Merge branch 'master' into sug-2-indicate-dirty-nodes-on-canvas

autologie committed:
fix: handle sub nodes properly

autologie committed:
change edge label color

autologie committed:
do not change dirtiness when node is enabled

autologie committed:
make immediate downstream node dirty when pinned data changed

autologie committed:
fix label color

autologie committed:
update dirtiness check for enabling sub node and removing a node

autologie committed:
Merge branch 'master' into sug-2-indicate-dirty-nodes-on-canvas

autologie committed:
fix failing unit tests

autologie committed:
add e2e test case

autologie committed:
fix typo

autologie committed:
replace dirty node icon

autologie committed:
fix: warning icon is missing in OutputPanel when there are multiple runData

autologie committed:
feat: mark start node of the loop dirty

autologie committed:
change the warning icon used in OutputPanel

autologie committed:
show run count next to the warning icon

autologie committed:
don't turn dirty node's outgoing edge yellow

autologie committed:
Merge branch 'master' into sug-2-indicate-dirty-nodes-on-canvas

autologie committed:
fix: removing connection should not make downstream node dirty

autologie committed:
update test to use realistic node ID

autologie committed:
fix: updating note and display-note option should not change dirtiness

autologie committed:
add a test case for unfixed bug

autologie committed:
perf: avoid getCurrentWorkflow() in calculating dirtiness

autologie committed:
update tests

autologie committed:
fix failing test

autologie commented:
## Human generated summary of changes since last approval

- After discussing with the product team, there are several changes to the conditions that makes a node dirty
  - We decided to not mark all downstream nodes dirty because it felt visually too much
  - In addition to node's parameter changes, following conditions also make a node dirty:
     1. Changes to incoming connections
     2. Updating and removing pinned data
     3. The first node in a loop where there is a dirty node
 - The logic for calculating dirtiness is extracted from `workflows.store.ts` to `useNodeDirtiness.ts`
 - Timestamp is added to undo/redo history items for dirtiness calculation
 - Added an e2e and more unit tests
 - Miscellaneous bugfixes and cosmetic changes

autologie committed:
fix loop detection

r00gm commented on packages/editor-ui/src/components/ResourceMapper/ResourceMapper.vue:
```diff
@@ -354,11 +355,14 @@ async function loadAndSetFieldsToMap(): Promise<void> {
-		emitValueChanged();
+
+		if (!isEqual(newSchema, state.paramValue.schema)) {
+			state.paramValue = {
+				...state.paramValue,
+				schema: newSchema,
+			};
```
Mutate the state instead of replacing it
```suggestion
			state.paramValue.schema = newSchema
```

r00gm commented on packages/editor-ui/src/components/canvas/elements/edges/CanvasEdge.vue:
```diff
@@ -87,7 +87,7 @@ const edgeClasses = computed(() => ({
 const edgeLabelStyle = computed(() => ({
 	transform: `translate(0, ${isConnectorStraight.value ? '-100%' : '0%'})`,
-	color: edgeColor.value,
+	color: 'var(--color-text-base)',
```
is this change intentional?

autologie replied:
Yes, we agreed to change the color for better visibility.

autologie committed:
refactor: use constant for dirtiness type

autologie committed:
refactor: replace usage of FontAwesomeIcon with N8nIcon

autologie committed:
refactor: avoid unnecessary object replacement

r00gm reviewed (approved):
very nice job 🚀

janober commented:
Got released with [`n8n@1.82.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.82.0)

netroy commented on packages/workflow/src/Workflow.ts:
```diff
@@ -143,7 +145,7 @@ export class Workflow {
 	 *
 	 */
-	__getConnectionsByDestination(connections: IConnections): IConnections {
+	static getConnectionsByDestination(connections: IConnections): IConnections {
```
what was the intention behind making this `static` ?
